### ISSUE TITLE
How to extract the nnUNet features

### ISSUE THREAD

**Issue by MehriB** [MIC-DKFZ/nnUNet#608]
Hi,

I am using nnUNet for segmentation of brain structures from MRI images and I am trying to see if I can somehow cluster my training images. As clustering features, I would like to use the nnUNet features. 
Is there a way to extract the last feature layer of the network before feeding it to the activation layer? So that I can use those features for my clustering?
Or is there a way to lock the activation in an intermediate layer of the prediction process?

Thank you for your help,
Mehri

**Comment by FabianIsensee** [comment]
Hi Mehri,
you will need to implement that yourself I am afraid. There is no baked-in functionality to give you these features.
Best,
Fabian

**Comment by MehriB** [comment]
Thank you Fabian for the reply.
Any tip on which part of the code I should be looking at?

Best,
Mehri

**Comment by WendyNice** [comment]
I want to get the features of encoder too. But don't know which part of code should I  change.

**Comment by FabianIsensee** [comment]
Go into the GenericUNet class and have a look at the forward function. You can modify it such that it stores or returns the encoder output

**Comment by Jun-electrophysiology** [comment]
> 

Hi,

I also wish to extract the feature map in the last layer of encoder. Could you please share your code?  
Many thanks!

Jun

**Comment by Jun-electrophysiology** [comment]
@WendyNice 

Hi,

I also wish to extract the feature map in the last layer of encoder. Could you please share your code?  
Many thanks!

Jun

**Comment by WendyNice** [comment]
I think my way to extract the feature map is a little complicated. I disable the cropping and mirroring operation, and than I changed the forward function in class Generic_UNet(SegmentationNetwork) in file network_architecture.generic_UNet.py

**Comment by WendyNice** [comment]
I think my way to extract the feature map is a little complicated. I disable the cropping and mirroring operation, and then I changed the forward function in class Generic_UNet(SegmentationNetwork) in file network_architecture.generic_UNet.py as below, so the forward will return the feature map but not seg_outputs

    def forward(self, x):
        # print('self.conv_blocks_context', len(self.conv_blocks_context), self.conv_blocks_context)
        # print('self.tu', len(self.tu), self.tu)
        skips = []
        seg_outputs = []
        for d in range(len(self.conv_blocks_context) - 1):
            x = self.conv_blocks_context[d](x)
            # print('feature map in', x.shape)
            skips.append(x)
            if not self.convolutional_pooling:
                x = self.td[d](x)

        x = self.conv_blocks_context[-1](x)
        # print('feature map', x.shape)
        # feature_map = x
        #
        # for u in range(len(self.tu)):
        #     x = self.tu[u](x)
        #     x = torch.cat((x, skips[-(u + 1)]), dim=1)
        #     x = self.conv_blocks_localization[u](x)
        #     print('x', x.shape)
        #     seg_outputs.append(self.final_nonlin(self.seg_outputs[u](x)))
        #
        # if self._deep_supervision and self.do_ds:
        #     return tuple([seg_outputs[-1]] + [i(j) for i, j in
        #                                       zip(list(self.upscale_logits_ops)[::-1], seg_outputs[:-1][::-1])])
        # else:
        #     print('run this')
        #     print('seg_outputs[-1]', seg_outputs[-1].shape)
        #     return seg_outputs[-1], x
        return x

**Comment by Jun-electrophysiology** [comment]
@WendyNice 
Thank you very much for sharing! I will try it.

**Comment by Jun-electrophysiology** [comment]
@WendyNice 
I have found a new problem. In my case, the patch size is 80*192*160, and the size of the last layer of encode is 1*320*5*6*5. However, the original image size is 160*384*384. How to get the whole original image's  feature map of the last layer of encode?

**Comment by WendyNice** [comment]
I solved this problem by disabling the cropping and sliding window 

1. change the crop_to_nonzero in preprocessing.cropping
change: 
bbox = get_bbox_from_mask(nonzero_mask, 0)
to: 
bbox = [[0, 155], [0, 225], [40, 200]]
(the size can be modified by your image size)

2. change the predict_cases in inference.predict.py
change:
        feature_map = trainer.predict_preprocessed_data_return_seg_and_softmax(
            d, do_mirroring=do_tta, mirror_axes=trainer.data_aug_params['mirror_axes'], use_sliding_window=True,
            step_size=step_size, use_gaussian=False, all_in_gpu=all_in_gpu,
            mixed_precision=mixed_precision)
to:
        feature_map = trainer.predict_preprocessed_data_return_seg_and_softmax(
            d, do_mirroring=False, mirror_axes=trainer.data_aug_params['mirror_axes'], use_sliding_window=False,
            step_size=step_size, use_gaussian=False, all_in_gpu=all_in_gpu,
            mixed_precision=mixed_precision)

**Comment by Jun-electrophysiology** [comment]
@WendyNice 
Thank you very, very much. You've really helped me a lot. Thank you very much!

**Comment by Yikeru** [comment]
@WendyNice 
Hi, am also trying to extract the feature maps. I basically adapted the code just as you have shown in their appropriate locations. I have some questions, so I apologize in advance if the answers to these questions seem very obvious/trivial. So I was wondering which function you call upon to basically get your feature maps.  I assume you just run the `nnUNet_predict` function in the terminal, with the basic outputs? If so where are you feature_maps exactly stored? Or did you make a seperate script for that?

**Comment by robsver** [comment]
Hi @WendyNice,
since the code has changed, is it still necessary to make these changes ? 
If so, do you have any idea what changes should be made in the new code because I don't find the following code:
```
feature_map = trainer.predict_preprocessed_data_return_seg_and_softmax(d, do_mirroring=False, mirror_axes=trainer.data_aug_params['mirror_axes'], use_sliding_window=False, step_size=step_size, use_gaussian=False, all_in_gpu=all_in_gpu, mixed_precision=mixed_precision)
```

Roberto

**Comment by pzaffino** [comment]
Hi,
I have the same question about the use_sliding_window parameter.

**Comment by MHD-s** [comment]
Hi @robsver,
